### ISSUE TITLE
ADAMA can pass specific headers along for passthrough services

### ISSUE THREAD

**Issue by mwvaughn** [Arabidopsis-Information-Portal/adama#88]
In [another discussion](https://github.com/Arabidopsis-Information-Portal/adama/commit/bf32ab8d856cca552c5010c40c94ac94975cafe6#commitcomment-13425833), it has been requested to pass headers along to remote services when creating a pass-through type service in Adama. This could cause weirdness if we just pass *everything* by default.  I would like to recommend instead that the service creator can request forwarding of specific headers when enrolling with Adama. This whitelist approach allows us, in turn, to specifically deny headers that we believe could lead to harmful outcomes. 

**Mockup for Headers.yml**

```yaml
---
forward:
    - 'x-jwt-assertion-araport-org'
    - 'user-agent'
deny:
    - 'authentication'
    - 'range'
```

**Comment by waltermoreira** [comment]
Similar to issue #44, arbitrarily passing the bearer token or JWT is a security issue, if the destination is not trusted.  I had in mind to add the option to allow specific headers.  Matt V's proposal looks good.

**Comment by mwvaughn** [comment]
Building on this, we should have an explicit blacklist that only trusted (in this case specific project members) users can override. That level of entitlement might be considered analogous to the admin role in agaveapi.

**Comment by mwvaughn** [comment]
I would add the Auth headers to that blacklist

**Comment by vivekkrish** [comment]
Hi @mwvaughn , I agree this a prudent decision to allow the developer of the adapter to control which particular set of headers they wish to pass-through to the destination.

Your mockup/propsal of a separate headers.yml file looks good.

Another option: since ADAMA reads the whitelist from `metadata.yml`, could it automatically decide to pass along the `X-Jwt-Assertion-Araport-Org` and `Authorization` headers to any services which have the  URL pattern `*.araport.org` in their whitelist?

**Comment by vivekkrish** [comment]
Following up on my suggestion above, it would avoid the need to have a separate config file, and in any case, the Token/JWT are only compatible with Araport services (Agave, ADAMA, ThaleMine, etc.)

**Comment by mwvaughn** [comment]
Vivek, good idea to be able to set a 'tenant'-level set of headers to pass along. However, we still want to also implement fine-grained controls. Let's look at this possible implementation instead:

```yaml
--- 
whitelist: 
  other.araport.org: 
    deny: 
      - authentication
      - origin
    forward: 
      - x-jwt-assertion-araport-org
      - user-agent
  iplant.plantontology.org: ~
  palea.cgrb.oregonstate.edu: 
    forward: 
      - range
```

The default for non tenant-hostname (aka not araport.org in our case) is ```deny: *```, while a tenant admin (Walter in this case) could set a policy for in-tenant hosts that was more permissive. However, in a case where araport.org was operating a service that we DIDN'T want to pass specific headers to, we still would have the ability to restrict them via how we configure the service. The approach of modifying the whitelist allows us to also offer this capability to non-passthrough services, permitting them to collect demographic fields, cookies, etc if needed.

**Comment by vivekkrish** [comment]
Hi Matt, I agree! Fine-grained controls coupled with tenant level set of headers seems like a good implementation. Your above proposal looks good.

Does this implementation require a separate `headers.yml` file or can this just be folded into the existing `metadata.yml` (since we already define the ``whitelist`` there)?

**Comment by mwvaughn** [comment]
^ this would live in a new version of the whitelist :sunglasses:

**Comment by eriksf** [comment]
Matt, I second the above proposal! I think it offers the flexibility to handle almost any situation but with sensible defaults.

**Comment by mwvaughn** [comment]
Cool, Walter and I will drop this on the roadmap. If it's easy, it'll come soon, and if not, we'll sort out a workaround to allow Thalemine to get the info it needs.

**Comment by waltermoreira** [comment]
A slight modification to the proposal: to make it backwards compatible, and to make it easy for the simple default case, I propose the syntax to be:

```
whitelist: 
  - iplant.plantontology.org
  - other.araport.org: 
      deny: 
        - authentication
        - origin
      forward: 
        - x-jwt-assertion-araport-org
        - user-agent
```

That is, `whitelist` is still a list, but its elements can be strings (which is the default case `deny: *`), or it can be an object whose key is the domain and the value is the forward policy.

What do people say?

**Comment by eriksf** [comment]
That works for me. I like that an old metadata.yml will work without modification in case the developer is unaware of the new syntax.

**Comment by vivekkrish** [comment]
Thats a great thought @waltermoreira ! Backwards compatibility of the `metadata.yml` and retaining uniformity when describing the `whitelist` as a list of URLs, will ensure that current developers won't be confused when they have to upgrade their adapters to have fine-grained support to forward along headers of their choice.

BTW, does the `whitelist` have to be absolute URLs or can it support patterns like `*.araport.org`? I'm assuming not, since it is used to set up the `iptables` of each worker container.

**Comment by waltermoreira** [comment]
@vivekkrish yes, elements of `whitelist` have to be complete domain names that resolve via dns, or ip's.  Do you have an use case for the wildcards?  I'm up for discussing it, but it's not clear how it would be implemented since the set of ip's could be dynamic in such case.

**Comment by mwvaughn** [comment]
I agree with Walter's proposed change to the data structure. Backwards compatibility and simplicity in one fell swoop.

**Comment by eriksf** [comment]
@mwvaughn & @waltermoreira , where does this fall on the priority list? If it's not high (which is totally fine), would it be possible to set up a workaround in the interim?

**Comment by waltermoreira** [comment]
@eriksf working on it right now. Expect something for tomorrow.

**Comment by eriksf** [comment]
Awesome! Thanks @waltermoreira !

**Comment by vivekkrish** [comment]
Wow, thats quick @waltermoreira , thank you!

Regarding the use of wildcards in the `whitelist`, there wasn't any particular use case that I was trying to address, it was more in reference to the origins of this discussion.  As I was typing out my last response to you, I realized why we need a full URL in the whitelist.

In conclusion, please ignore my question about supporting wildcards. ;)

Look forward to seeing the above proposed changes in ADAMA soon. Will this be deployed first to adama-dev?

**Comment by mwvaughn** [comment]
In case it breaks things, let's do put it on ADAMA-dev first :8ball:

**Comment by waltermoreira** [comment]
Unfortunately, we don't have Azure anymore, that's where the `adama-dev.cloudapp.net` was hosted.  I usually test in my own vagrant boxes, and today I created a rodeo instance, but it doesn't have any domain name pointing at it.  I'll push it there and send the ip (assuming rodeo keeps working for more than one day...).

**Comment by waltermoreira** [comment]
Just created a dev instance in AWS and pointed `adama-dev.waltermoreira.net` to it (sorry for the personal domain, but it's faster than opening tickets requiring dns changes to the powers that be).  I'll send you a token to access it:

```
$ http http://adama-dev.waltermoreira.net/community/v0.3/status
...
    "hash": "d90670f2b56810211c772274a1e739d26bcc3a38",
...
```

There is a sample service with the new whitelist features at https://github.com/waltermoreira/httpbin-test

Please, give it a test.  If things work, pushing to production should be fast as it doesn't require rebuilding the existing services.

**Comment by eriksf** [comment]
@waltermoreira, thanks for the quick turnaround!

I made a clone of your httpbin-test repo, modified it, and gave it a shot on the dev instance:

```
---

name: httpbin_get
type: passthrough
url: http://httpbin.org/get
whitelist:
  - httpbin.org:
      deny:
        - foo
        - bar
      forward:
        - baz
        - X-Jwt-Assertion-Araport-Org
        - Authorization
  - example.com
```

It seems to work great. Any headers in deny don't get passed and anything in forward do. I can't truly test because I'm not getting a valid JWT, but it's passing X-Jwt-Assertion-Araport-Org along when I set it. Ultimately, I don't really need the Authorization header but it works anyway.

```
$ http -v http://adama-dev.waltermoreira.net/community/v0.3/erik/httpbin_get_v0.1/access foo:one bar:two baz:three X-Jwt-Assertion-Araport-Org:erik
GET /community/v0.3/erik/httpbin_get_v0.1/access HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Authorization: Bearer xxxxx
Connection: keep-alive
Host: adama-dev.waltermoreira.net
User-Agent: HTTPie/0.9.2
X-Jwt-Assertion-Araport-Org: erik
bar: two
baz: three
foo: one



HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Connection: keep-alive
Content-Length: 405
Content-Type: application/json
Link: http://adama-dev.waltermoreira.net/community/v0.3/erik/httpbin_get_v0.1/prov/6e6f163e4cbb4e58b0209415919c57d8; rel="http://www.w3.org/ns/prov#has_provenance"
access-control-allow-credentials: true
date: Mon, 28 Sep 2015 23:16:22 GMT
server: nginx

{
    "args": {}, 
    "headers": {
        "Accept": "*/*", 
        "Accept-Encoding": "gzip, deflate", 
        "Authorization": "Bearer xxxxx", 
        "Baz": "three", 
        "Host": "httpbin.org", 
        "User-Agent": "python-requests/2.7.0 CPython/2.7.6 Linux/3.13.0-48-generic", 
        "X-Jwt-Assertion-Araport-Org": "erik"
    }, 
    "origin": "52.89.249.129", 
    "url": "http://httpbin.org/get"
}
```

**Comment by waltermoreira** [comment]
Has anyone tested a non-passthrough service in this instance?  If someone can publish a couple of different services just to test backward compatibility, then we are ready to deploy this change to production.

The test suite is passing in my development instance, so I'm pretty confident, but I still would like someone else trying.

Thanks!

**Comment by mwvaughn** [comment]
I'll try to do it 10/2/2015

**Comment by mwvaughn** [comment]
Confirmed: Query service ```aip/expressologs_by_locus_v0.2.0``` works as intended when registered at ```adama-dev```. This code is ready for prime time.

**Comment by waltermoreira** [comment]
Awesome, thanks, Matt!  I'll proceed with moving it to production.

**Comment by waltermoreira** [comment]
Fixed in commit d90670f2b56810211c772274a1e739d26bcc3a38 and deployed.
